### PR TITLE
cache results of fileExists check in default compiler host

### PR DESCRIPTION
**Pull request by vladima**

### CODE REVIEW

**Comment by RyanCavanaugh** [comment]
Isn't this a problem for `tsc -w` ?

**Comment by vladima** [comment]
it might. Also we currently already cache results of `directoryExists` check which is also incorrect

**Comment by RyanCavanaugh** [comment]
`directoryExists` does the right thing for that scenario, though -- it only catches `true` results

**Comment by vladima** [comment]
directory can be deleted between recompilations and we still will think it exists

**Comment by DanielRosenwasser** [comment]
You could just make a helper function for `sys.fileExists` that performs its own caching in non-watch scenarios.

File: src/compiler/tsc.ts [hunk 1]
```diff
@@ -286,6 +295,13 @@ namespace ts {
             reportWatchDiagnostic(createCompilerDiagnostic(Diagnostics.Compilation_complete_Watching_for_file_changes));
         }
 
+        function fileExists(fileName: string): boolean {
```
Comment (mhegazy) [inline src/compiler/tsc.ts]: can we rename this to `cachedFileExist` just to make it clear.

File: src/compiler/tsc.ts [hunk 2]
```diff
@@ -159,6 +159,9 @@ namespace ts {
         let timerHandleForRecompilation: number;                    // Handle for 0.25s wait timer to trigger recompilation
         let timerHandleForDirectoryChanges: number;                 // Handle for 0.25s wait timer to trigger directory change handler
 
+        let cachedExistingFiles: Map<boolean>;
```
Comment (mhegazy) [inline src/compiler/tsc.ts]: can you add a comment here to explain why this is needed.

**Comment by mhegazy** [comment]
:+1:

File: src/compiler/tsc.ts (hunk 1 above)
Comment (vladima) [inline src/compiler/tsc.ts]: sure

**Pull request merged**